rasa/shared/core/domain.py

```diff
@@ -252,195 +255,213 @@ def from_directory(cls, path: Text) -> "Domain":
-            combined[key] = self.merge_lists(combined[key], domain_dict[key])
+            duplicates[key] = rasa.shared.utils.common.extract_duplicates(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
+            combined[key] = rasa.shared.utils.common.merge_lists(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
 
-        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
-            combined[key] = self.merge_dicts(combined[key], domain_dict[key], override)
+        return combined, duplicates
 
-        return self.__class__.from_dict(combined)
+    @staticmethod
+    def _merge_forms_responses_slots(
+        combined: Dict, domain_dict: Dict, duplicates: Dict, override: bool = False
+    ) -> Tuple:
+        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
+            duplicates[key] = rasa.shared.utils.common.extract_duplicates(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
+            combined[key] = rasa.shared.utils.common.merge_dicts(
+                combined.get(key, {}), domain_dict.get(key, {}), override
+            )
+        return combined, duplicates
 
+    @staticmethod
     def merge_domain_dicts(
-        self, domain1: Dict, domain2: Dict, override: bool = False
-    ) -> Dict[Text, Any]:
-        """Merges this domain dict with another one, combining their attributes.
+        domain_dict: Dict, combined: Dict, override: bool = False, is_dir: bool = False
+    ) -> Dict:
+        """Combines two domain dictionaries."""
+        if not domain_dict:
+            return combined
 
-        This is used when multiple domain yml files are configured in a single
-        directory. Unlike the merge method above, which merges Domain objects by
-        creating each object then merging it with the previous, this method merges
-        domain dicts, and ensures all attributes (like ``intents``, ``entities``, and
-        ``actions``) are known to the Domain when the object is created.
+        if not combined:
+            return domain_dict
 
-        List attributes like ``intents`` and ``actions`` are deduped
-        and merged. Single attributes are taken from `domain1` unless
-        override is `True`, in which case they are taken from `domain2`.
-        """
-        if not domain2:
-            return domain1
-
-        if not domain1:
-            return domain2
-
-        domain_dict = domain2
-        combined = domain1
+        if not combined.get(KEY_TRAINING_DATA_FORMAT_VERSION):
+            combined[
+                KEY_TRAINING_DATA_FORMAT_VERSION
+            ] = LATEST_TRAINING_DATA_FORMAT_VERSION
 
         if override:
             config = domain_dict["config"]
             for key, val in config.items():
                 combined["config"][key] = val
 
-        if override or domain2.get("session_config"):
+        if (
+            override
+            or (
+                not is_dir
```

Comment:
Why do we care about `is_dir` here?

Reply:
In the original code, the 2 merge functions for loading a file and loading a directory differed in the 2nd part of this check `if override or domain2.get / combined.get` so in order to keep this difference i had to introduce this boolean `is_dir`.

Reply:
But what is the behavioural reason? I.e. what does this actually accomplish?

Reply:
Refactored to make it clear what is the desired behaviour for overriding SESSION_CONFIG_KEY with the key from the domain that gets merged.